### PR TITLE
Lockless diagnostics

### PR DESCRIPTION
**Pull request by pepeiborra**
Fourth instalment of #2357 

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/2434"></a>

### CODE REVIEW

File: ghcide/src/Development/IDE/Core/Shake.hs [hunk 1]
```diff
@@ -758,13 +758,11 @@ instantiateDelayedAction (DelayedAction _ s p a) = do
 
 getDiagnostics :: IdeState -> IO [FileDiagnostic]
 getDiagnostics IdeState{shakeExtras = ShakeExtras{diagnostics}} = do
-    val <- readVar diagnostics
-    return $ getAllDiagnostics val
+    atomically $ getAllDiagnostics diagnostics
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/Core/Shake.hs:761]: push to call-sites?

File: ghcide/src/Development/IDE/Core/Shake.hs [hunk 2]
```diff
@@ -758,13 +758,11 @@ instantiateDelayedAction (DelayedAction _ s p a) = do
...
 
 getHiddenDiagnostics :: IdeState -> IO [FileDiagnostic]
 getHiddenDiagnostics IdeState{shakeExtras = ShakeExtras{hiddenDiagnostics}} = do
-    val <- readVar hiddenDiagnostics
-    return $ getAllDiagnostics val
+    atomically $ getAllDiagnostics hiddenDiagnostics
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/Core/Shake.hs:765]: push to call sites?

File: ghcide/src/Development/IDE/Core/Shake.hs [hunk 3]
```diff
@@ -1154,30 +1152,26 @@ updateFileDiagnostics fp k ShakeExtras{logger, diagnostics, hiddenDiagnostics, p
...
+        update new store = setStageDiagnostics uri ver (T.pack $ show k) new store
     mask_ $ do
         -- Mask async exceptions to ensure that updated diagnostics are always
         -- published. Otherwise, we might never publish certain diagnostics if
         -- an exception strikes between modifyVar but before
         -- publishDiagnosticsNotification.
-        newDiags <- modifyVar diagnostics $ pure . update (map snd currentShown)
-        _ <- modifyVar hiddenDiagnostics $  pure . update (map snd currentHidden)
+        newDiags <- liftIO $ atomically $ update (map snd currentShown) diagnostics
+        _ <- liftIO $ atomically $ update (map snd currentHidden) hiddenDiagnostics
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/Core/Shake.hs:1162]: should these be in one transaction?

File: ghcide/src/Development/IDE/Core/Shake.hs [hunk 4]
```diff
@@ -191,9 +191,9 @@ data ShakeExtras = ShakeExtras
     ,logger :: Logger
     ,globals :: Var (HMap.HashMap TypeRep Dynamic)
     ,state :: Values
-    ,diagnostics :: Var DiagnosticStore
-    ,hiddenDiagnostics :: Var DiagnosticStore
-    ,publishedDiagnostics :: Var (HMap.HashMap NormalizedUri [Diagnostic])
+    ,diagnostics :: STMDiagnosticStore
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/Core/Shake.hs:194]: Is the `DiagnosticStore` type now dead?

**Review by michaelpj — APPROVED** [review]

File: ghcide/src/Development/IDE/Core/Shake.hs (hunk 4 above)
Comment (pepeiborra) [reply ghcide/src/Development/IDE/Core/Shake.hs:194]: I think it comes from the `lsp` package

File: ghcide/src/Development/IDE/Core/Shake.hs (hunk 3 above)
Comment (pepeiborra) [reply ghcide/src/Development/IDE/Core/Shake.hs:1163]: I don't believe so (a single transaction will introduce more contention in exchange for no more correctness)

Comment (pepeiborra) [reply ghcide/src/Development/IDE/Core/Shake.hs:1163]: notice that this block is already `mask`ed

File: ghcide/src/Development/IDE/Core/Shake.hs [hunk 5]
```diff
@@ -447,10 +447,11 @@ recordDirtyKeys
...
   -> k
   -> [NormalizedFilePath]
-  -> IO ()
-recordDirtyKeys ShakeExtras{dirtyKeys} key file = withEventTrace "recordDirtyKeys" $ \addEvent -> do
-    atomically $ modifyTVar' dirtyKeys $ \x -> foldl' (flip HSet.insert) x (toKey key <$> file)
-    addEvent (fromString $ "dirty " <> show key) (fromString $ unlines $ map fromNormalizedFilePath file)
+  -> STM (IO ())
+recordDirtyKeys ShakeExtras{dirtyKeys} key file = do
+    modifyTVar' dirtyKeys $ \x -> foldl' (flip HSet.insert) x (toKey key <$> file)
+    return $ withEventTrace "recordDirtyKeys" $ \addEvent -> do
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/Core/Shake.hs:453]: Hmm interesting. I guess we may end up with this happening a bit if we push STM out further and we hit more places where we want to do tracing. I wonder if this is a pattern that appears elsewhere: this is something like a "final" IO action that should be run only after the STM action succeeds.

File: ghcide/session-loader/Development/IDE/Session.hs [hunk 6]
```diff
@@ -247,7 +247,7 @@ loadSessionWithOptions SessionLoadingOptions{..} dir = do
                       } <- getShakeExtras
     let invalidateShakeCache = do
             void $ modifyVar' version succ
-            recordDirtyKeys extras GhcSessionIO [emptyFilePath]
+            atomically $ recordDirtyKeys extras GhcSessionIO [emptyFilePath]
```
Comment (fendor) [inline ghcide/session-loader/Development/IDE/Session.hs:250]: I just noticed, this changes the type of `invalidateShakeCache` to `:: IO (IO ())`. I suppose that is unintentional?

Comment (fendor) [reply ghcide/session-loader/Development/IDE/Session.hs:250]: I think, either here we need a `join` or have to consume the action produced at: https://github.com/haskell/haskell-language-server/pull/2434/files#diff-d7140bbfca1b7aaafd8c82d1d6aa21a3f941eb27ba8f08e4d9fb13766b21b418L393

**Pull request merged**